packages/webpack-plugin/src/plugin.ts

```diff
@@ -0,0 +1,227 @@
+          raw: true,
+          include: /\.(js|ts|jsx|tsx|mjs|cjs)(\?[^?]*)?(#[^#]*)?$/,
+          banner: (arg?: BannerPluginCallbackArg) => {
+            const hash = arg?.chunk?.contentHash?.javascript ?? arg?.chunk?.hash;
+            const debugId = hash ? stringToUUID(hash) : uuidv4();
+            return getDebugIdSnippet(debugId);
+          },
+        })
+      );
+    },
+  });
+}
+
+function webpackDebugIdUploadPlugin(): (
+  upload: (buildArtifacts: string[]) => Promise<void>,
+  logger: Logger,
+  forceExitOnBuildCompletion?: boolean
+) => UnpluginOptions {
+  const pluginName = "sentry-webpack-debug-id-upload-plugin";
+  return (
+    upload: (buildArtifacts: string[]) => Promise<void>,
+    logger: Logger,
+    forceExitOnBuildCompletion?: boolean
+  ) => ({
+    name: pluginName,
+    webpack(compiler) {
+      compiler.hooks.afterEmit.tapAsync(pluginName, (compilation, callback: () => void) => {
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+        const outputPath = (compilation.outputOptions.path as string | undefined) ?? path.resolve();
+        const buildArtifacts = Object.keys(compilation.assets as Record<string, unknown>).map(
+          (asset) => path.join(outputPath, asset)
+        );
+        void upload(buildArtifacts).then(() => {
+          callback();
+        });
+      });
+
+      if (forceExitOnBuildCompletion && compiler.options.mode === "production") {
+        compiler.hooks.done.tap(pluginName, () => {
+          setTimeout(() => {
+            logger.debug("Exiting process after debug file upload");
+            process.exit(0);
+          });
+        });
+      }
+    },
+  });
+}
+
+function webpackModuleMetadataInjectionPlugin(
+  UnsafeBannerPlugin: UnsafeBannerPlugin | undefined
+): (injectionCode: string) => UnpluginOptions {
+  return (injectionCode: string) => ({
+    name: "sentry-webpack-module-metadata-injection-plugin",
+    webpack(compiler) {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore webpack version compatibility shenanigans
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
+      const BannerPlugin =
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore webpack version compatibility shenanigans
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+        compiler?.webpack?.BannerPlugin || UnsafeBannerPlugin;
+
+      compiler.options.plugins = compiler.options.plugins || [];
+      compiler.options.plugins.push(
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-call
+        new BannerPlugin({
+          raw: true,
+          include: /\.(js|ts|jsx|tsx|mjs|cjs)(\?[^?]*)?(#[^#]*)?$/,
+          banner: injectionCode,
+        })
+      );
+    },
+  });
+}
+
+export function sentryWebpackUnpluginFactory({
```

Comment:
Let's add a comment here why this takes the banner/define plugins. ie: To not have to actually depend on webpack.